### PR TITLE
Add a provisional section on pytest code style w.r.t. fixtures

### PR DESCRIPTION
**Pull request by vshampor**
### Changes
As stated in the title

### Reason for changes
This doesn't seem obvious to some developers, so will state this in the style guide.

### Related tickets
N/A

### Tests
N/A

### CODE REVIEW

**Review by alexsu52 — APPROVED** [review]
LGTM

File: docs/styleguide/PyGuide.md [hunk 1]
```diff
@@ -816,3 +816,18 @@ This is required so that the autogenerated API documentation is rendered properl
...
+
+<a id="s5-test-suite-coding-rules"></a>
+<a id="5-test-suite-coding-rules"></a>
+<a id="test-suite-coding-rules"></a>
+
+## 5 Test suite coding rules
+
+Unit tests are written and executed using the [pytest](https://pytest.org) framework.
+
+*to_be_extended_with_more_rules_capturing_the_current_style*
```
Comment (AlexanderDokuchaev) [inline docs/styleguide/PyGuide.md:828]: Some suggestion to this:
- use pathlib.Path for fixtures for paths. Now it's all return str
- Add fixtures to scripts wehere it's used, and avoid add to conftest.py only it's common fixtures that can used in many tests.
- tests does not contain logic to manual comparing results as current e2e. 
- Used docstring and type hints for any fixtures
- Add best practics for rare case like colleting data from the tests ( like https://github.com/openvinotoolkit/nncf/pull/2152#discussion_r1373915533) 
- Add bad practics like this https://github.com/openvinotoolkit/nncf/blob/76bb1f085ef21fe000ba3726115bf4c84d6ac937/tests/tensorflow/conftest.py#L73 


May be you can analyze our tests to provide more examples of good and bad practice.

File: docs/styleguide/PyGuide.md [hunk 2]
```diff
@@ -816,3 +816,18 @@ This is required so that the autogenerated API documentation is rendered properl
 
 If the autogenerated API documentation does not show type hints for certain arguments despite the fact that the type hints are present in the object's implementation code,
 or if the type hints do not refer to the API symbol's canonical alias, then the type hint should be explicitly declared in the docstring using the `:type *param_name*:` directive (or `:rtype:` for return types).
+
+<a id="s5-test-suite-coding-rules"></a>
+<a id="5-test-suite-coding-rules"></a>
+<a id="test-suite-coding-rules"></a>
+
+## 5 Test suite coding rules
```
Comment (AlexanderDokuchaev) [inline docs/styleguide/PyGuide.md:824]: Should be index 6. 
And link to "Table of Contents"

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: docs/styleguide/PyGuide.md (hunk 1 above)
Comment (AlexanderDokuchaev) [reply docs/styleguide/PyGuide.md:828]: And will be good add point about add point about temples that used that used for PTQ.
My suggestion to add all templates to `tests/shared/test_templates`.

File: docs/styleguide/PyGuide.md (hunk 2 above)
Comment (vshampor) [reply docs/styleguide/PyGuide.md:824]: I would expect this to be caught by markdownlint. Thanks for spotting.

File: docs/styleguide/PyGuide.md (hunk 1 above)
Comment (vshampor) [reply docs/styleguide/PyGuide.md]: See the updated version. The global variable usage with `pytest.metrics_dump_path` is covered by the guide in section 3.3. Added the pathlib usage to section 3.7. `tests/shared` is actually meant for non-test-case code reused in other test folders, I meant for the actual test code that would be reused across multiple frameworks to be located in `tests/cross_fw`

File: docs/styleguide/PyGuide.md (hunk 2 above)
Comment (AlexanderDokuchaev) [reply docs/styleguide/PyGuide.md]: unfortunately markdownlint doesn't check index in headers

File: docs/styleguide/PyGuide.md [hunk 3]
```diff
@@ -816,3 +820,74 @@ This is required so that the autogenerated API documentation is rendered properl
...
+Good:
+
+```python3
+@dataclass
+class StructForTest:
+  expected_return_value: int
+  prohibited_attribute_value: str
+  raises: Optional[Type[Exception]]
+  
+@pytest.mark.parametrize([StructForTest(expected_return_value=42,
```
Comment (AlexanderDokuchaev) [inline docs/styleguide/PyGuide.md:866]: Missed name of parameter
```python 
@pytest.mark.parametrize(
    "ts",  # <----
    [
        StructForTest(expected_return_value=42, prohibited_attribute_value="foo", raises=None),
        StructForTest(expected_return_value=1337, prohibited_attribute_value="bar", raises=RuntimeError),
    ],
)
```

Please provide code after black formatting.
And looks like need `idfn` for this case.

File: docs/styleguide/PyGuide.md [hunk 4]
```diff
@@ -816,3 +820,74 @@ This is required so that the autogenerated API documentation is rendered properl
...
+[Fixture scoping](https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session) should be leveraged depending on the use case - for instance, a class-scoped fixture can be utilized to provide a storage object for data gathered in individual cases inside the class, and to post-process this data after the test cases in the class have all been executed.
+The fixtures that are meant to be reused in multiple test files should be defined in a corresponding `conftest.py` file, instead of importing these where they are needed from another non-`conftest.py` file where they were defined - otherwise the linters will try to remove it as an unused import and also the import itself may have an effect on how the fixture is executed based on its scope.
+
+Test cases may be parametrized either using a `@pytest.mark.parametrize` decorator (for simple parametrizations), or by using [parametrized fixtures](https://docs.pytest.org/en/6.2.x/fixture.html#parametrizing-fixtures) in case the test requires complex parametrization, or in case the parameters have to be preprocessed in a complex fashion before being passed into the test.
```
Comment (AlexanderDokuchaev) [inline docs/styleguide/PyGuide.md:839]: Why used 6.2 version of documentation?

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: docs/styleguide/PyGuide.md (hunk 1 above)
Comment (AlexanderDokuchaev) [reply docs/styleguide/PyGuide.md]: Looks good 👍 
About templates is ok, i will replace templates.

File: docs/styleguide/PyGuide.md (hunk 4 above)
Comment (vshampor) [reply docs/styleguide/PyGuide.md:839]: Changed to 7.4 as in the rest of the links

File: docs/styleguide/PyGuide.md (hunk 3 above)
Comment (vshampor) [reply docs/styleguide/PyGuide.md:866]: Done

File: docs/styleguide/PyGuide.md (hunk 4 above)
Comment (AlexanderDokuchaev) [reply docs/styleguide/PyGuide.md]: But now it's not valid link, `Parametrizing fixtures` does not exist in 7.4

Comment (vshampor) [reply docs/styleguide/PyGuide.md]: Fixed

File: docs/styleguide/PyGuide.md [hunk 5]
```diff
@@ -794,3 +798,87 @@ This is required so that the autogenerated API documentation is rendered properl
...
+
+<a id="s6-test-suite-coding-rules"></a>
+<a id="6-test-suite-coding-rules"></a>
+<a id="test-suite-coding-rules"></a>
+
+## 6 Test suite coding rules
+
+Unit tests are written and executed using the [pytest](https://pytest.org) framework.
+
+Parameters of test cases and fixtures should be type-hinted just like any other function.
```
Comment (AlexanderDokuchaev) [inline docs/styleguide/PyGuide.md:810]: Split rules by headers like 
```
### Fixture
...
```

And will be good to add rule that the module versions should be determined in the test requirements.
```
torch  # bad 
torch == 2.1.0  # good
```

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: docs/styleguide/PyGuide.md (hunk 5 above)
Comment (vshampor) [reply docs/styleguide/PyGuide.md:810]: Done

Comment (AlexanderDokuchaev) [reply docs/styleguide/PyGuide.md:810]: Done what? Last commit is https://github.com/openvinotoolkit/nncf/pull/2220/commits/c84451794c24b64b2d86280fde629af16832fb33

Comment (vshampor) [reply docs/styleguide/PyGuide.md:823]: Pushed the new commit

**Review by AlexanderDokuchaev — APPROVED** [review]

**Pull request merged**